### PR TITLE
Close #172

### LINKED ISSUE #172: Implement EqualsAndGetHashcodeNotImplementedTogether
When either of the two methods are implemented, the other one has to be as well. I would suggest to add a codefix which implements the method signature and throws a `NotImplementedException()` from the body.

### PR DESCRIPTION
**Pull request by Hosch250**
Cloes #172

### CODE REVIEW

File: VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                }
+
+                var methodDeclaration = (MethodDeclarationSyntax)node;
+
+                if (!IsOverride(methodDeclaration))
+                {
+                    continue;
+                }
+
+                if (methodDeclaration.Identifier.ValueText == nameof(Equals) && methodDeclaration.ParameterList.Parameters.Count == 1)
```
Comment (Vannevelj) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs]: I don't like this approach. However there is a more solid alternative: you can request the symbol of `Object+Equals(object)` with something like this:

	var objectSymbol = compilation.GetSpecialType(SpecialType.System_Object);
	var objectEquals = objectSymbol.GetMembers().OfType<IMethodSymbol>().Where(x => x.MetadataName == "Equals" && x.Parameters.Count() == 1);

Afterwards you can use `context.SemanticModel.GetDeclaredSymbol(methodDeclaration).OverriddenMethod` and compare it to the the `Object+Equals(object)` method. 

Some things to note:

* Fetch `Object+Equals(object)` only once per compilation. You can do this by registering a compilation action in `Initialize()`
* You have to iterate through the inheritance structure in case a parent class already overrides it

File: VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            }
+
+            if (equalsImplemented ^ getHashcodeImplemented)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(Rule, classDeclaration.GetLocation(),
+                    ImmutableDictionary.CreateRange(new[] { new KeyValuePair<string, string>("IsEqualsImplemented", equalsImplemented.ToString()) })));
+            }
+        }
+
+        private bool IsOverride(MethodDeclarationSyntax methodDeclaration)
```
Comment (Vannevelj) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs]: This is overzealous with LINQ: a `Contains()` check is perfectly fine (it doesn't allocate anything).

File: VSDiagnostics/VSDiagnostics/VSDiagnostics/VSDiagnosticsResources.resx [hunk 3]
```diff
@@ -509,4 +509,13 @@
   <data name="SwitchDoesNotHandleAllEnumOptionsCodeFixTitle" xml:space="preserve">
     <value>Add case(s).</value>
   </data>
+  <data name="EqualsAndGetHashcodeNotImplementedTogetherAnalyzerMessage" xml:space="preserve">
+    <value>Both Equals() and GetHashcode() must be implemented.</value>
```
Comment (Vannevelj) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/VSDiagnosticsResources.resx]: I would reword this to "Equals() and GetHashcode() must be implemented together" to stress that that discrepancy is the problem

File: VSDiagnostics/VSDiagnostics/VSDiagnostics/VSDiagnosticsResources.resx [hunk 4]
```diff
@@ -509,4 +509,13 @@
...
+  </data>
+  <data name="EqualsAndGetHashcodeNotImplementedTogetherAnalyzerTitle" xml:space="preserve">
+    <value>Implement Equals() and GetHashcode().</value>
```
Comment (Vannevelj) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/VSDiagnosticsResources.resx]: That's the message for a different analyzer (the one that just provides a context action). Instead, I would make it "Implement Equals() and GetHashcode() together"

File: VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherCodeFix.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var diagnostic = context.Diagnostics.First();
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+
+            var statement = root.FindNode(diagnosticSpan);
+
+            if (bool.Parse(diagnostic.Properties["IsEqualsImplemented"]))
```
Comment (Vannevelj) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherCodeFix.cs]: You can compress this `if` by moving the entire `context.RegisterCodeFix()` right outside of it and just taking care of the code fix title inside the if (perhaps even by using a ternary if).

Something like this:

	var message = bool.Parse(diagnostic.Properties["IsEqualsImplemented"])
					? "GetHashCode()"
					: "Equals()";

	context.RegisterCodeFix(
		 CodeAction.Create(string.Format(VSDiagnosticsResources.EqualsAndGetHashcodeNotImplementedTogetherCodeFixTitle, message),
		 x => ImplementGetHashCodeAsync(context.Document, root, statement),
		 EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.Rule.Id),
		 diagnostic);

**Comment by Vannevelj** [comment]
Additionally: can you update `README.md` to include the new analyzer?

File: VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs (hunk 2 above)
Comment (Hosch250) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs]: This isn't a `Contains` call,  the Linq version would be `return methodDeclaration.Modifiers.Any(modifier => modifier.IsKind(SyntaxKind.OverrideKeyword));`

**Comment by Hosch250** [comment]
I'll update the Readme when it gets merged.  I think there are a couple others missing as well.

File: VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs [hunk 6]
```diff
@@ -39,18 +45,17 @@ private void AnalyzeNode(SyntaxNodeAnalysisContext context)
...
 
                 var methodDeclaration = (MethodDeclarationSyntax)node;
-
                 if (!IsOverride(methodDeclaration))
                 {
                     continue;
                 }
 
-                if (methodDeclaration.Identifier.ValueText == nameof(Equals) && methodDeclaration.ParameterList.Parameters.Count == 1)
+                if (context.SemanticModel.GetDeclaredSymbol(methodDeclaration).OverriddenMethod == objectEquals)
```
Comment (Vannevelj) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs]: What does `OverriddenMethod` refer to? Will this still work if you have a class `X : Object` that implements both methods but also have `Y : X` which violates the rule? Can you add a test to make sure this works?

Comment (Hosch250) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs]: This case does not work.  For the base class, the containing namespace of the parent Equals/GetHashCode is System.Object.  For the derived class, the parent namespace is the namespace X is defined in.  This causes the expected and actual Equals/GetHashCode methods to be different.

Comment (Hosch250) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs]: Of course, I could recurse the `OverriddenMethod` value to see if we ever reach the expected override on `Object`.  That should solve the problem.  To answer your question, `OverriddenMethod` refers to the method the current method is overridding.

Comment (Vannevelj) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs]: Yeah, that was the solution I was alluding to. I think you've already added a commit that fixes this now

Comment (Hosch250) [inline VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/EqualsAndGetHashcodeNotImplementedTogether/EqualsAndGetHashcodeNotImplementedTogetherAnalyzer.cs]: Yep, that was #4aec46c
